### ISSUE TITLE
Access to epsilons from traininglistener

### ISSUE THREAD

**Issue by DrChainsaw** [deeplearning4j/deeplearning4j#6494]
I'm trying to implement the pruning method from this paper:
https://arxiv.org/abs/1611.06440

Short summary is that a very good metric for puning neurons is abs(mean(epsilon * activation)) (with reservation for me misreading the paper :) )

As far as I can tell, the epsilons are not accessible through the training listener callback. 

As a workaround I have made an EpsilonSpyVertex but it is a bit of a pita to insert it at the right place and then remove when not training as extra layers/vertexes (even no-ops) seem to add processing latency (presumably due to workspace leveraging)

**Comment by AlexDBlack** [comment]
That's correct. Currently the epsilons (i.e., activation gradients, dL/daOut with aOut being the activations of a layer) are not really made available.
I haven't read the paper, but I assume you would want them for all layers?
One challenge in providing them in a listener API is that for performance and memory reasons, they are present in workspaces. i.e., they no longer exist for most layers by the time the backprop is complete (their memory has already been reused).

One hack I can think of for this is to add a preprocessor on every layer that simply duplicates/detaches the epsilons and stores them somewhere. It would work, but it would cost you a fair bit of memory/performance - which might be fine for your use case.

**Comment by DrChainsaw** [comment]
Thanks @AlexDBlack. I guess this is a kind of an odd case so I'll keep using the hack. Didn't think of the preprocessor so I just made a new GraphVertex. Would there be any benefits of a preprocessor instead (except for maybe on less class)? I guess I can't attach them to a built graph so I would still need to insert them at build time.

Stuff below is perhaps a bit besides the "issue" itself, but related. Answer if you have the time for it :)

I run oom in just a few iterations if I don't put the arrays in a workspace. You don't happen to have any recommendations for a config? Here is the one I currently use (one per layer as I have one listener per layer):
```
private final WorkspaceConfiguration workspaceConfig = WorkspaceConfiguration.builder()
                .policyAllocation(AllocationPolicy.STRICT) // Because the size is not gonna change?
                .policyLearning(LearningPolicy.OVER_TIME) // To make cyclesBeforeInitialization take effect?
                .cyclesBeforeInitialization(2) // Because after two arrays the size should be fully known?
                .policyReset(ResetPolicy.ENDOFBUFFER_REACHED) // To make it cyclic?
                .policySpill(SpillPolicy.REALLOCATE)  // Does not matter?
                .initialSize(0) // Any benefit over time with > 0?
                .build(); 
```
It does the job alot better but program still leaks memory somehow and eventually crashes, but that might have nothing to do with the workspace. 

On a more crazy note: I do get a significant slowdown, especially if I put the gleaned arrays in a HOST_ONLY workspace to conserve GPU memory. The program becomes about 200-800% slower per batch.

Migrating the arrays and doing the computation is kinda low prio background work. In theory I could run it in a separate thread and even skip/discard data to make sure I don't block the training. I guess I would be in a world of pain if I tried to realize something like this when running with the CUDA backend. Is it possible to use some CPU-only OP to work on the arrays so that I could put it in a separate thread (I would of course then need to invent some way to drop/skip such jobs once they become blocking)?

**Comment by AlexDBlack** [comment]
> Would there be any benefits of a preprocessor instead (except for maybe on less class)? I guess I can't attach them to a built graph so I would still need to insert them at build time.

Preprocessors would be easier to insert into the config (in the layer vertices config before before you instantiate the ComputationGraph). Or yes, you have to build the net from scratch with them attached.

> I run oom in just a few iterations if I don't put the arrays in a workspace.

That's not surprising, activation gradients are very large in some nets. Alternatively, you can decrease the garbage collection frequency (i.e., more frequent) - or even write to disk. Or decrease the minibatch size.

> program still leaks memory somehow and eventually crashes

We've seen very stable memory use somewhere. Make sure you're not keeping array references around any more than you need.

> I do get a significant slowdown, especially if I put the gleaned arrays in a HOST_ONLY workspace to conserve GPU memory.

Well, yeah. Copying many GB of memory back to the host has it's cost :)

> Is it possible to use some CPU-only OP to work on the arrays

In practice - no, we don't support this in ND4J (yet). In principle, sure.

**Comment by DrChainsaw** [comment]
@AlexDBlack 
Thanks for input. Really appreciate it!

Sorry to only reward you with even more questions:

One thing I don't understand is this: After training for one epoch, I must change the name of the workspace or else I get some kind of low level fault (iirc it was cudastream sync failed or EXCEPTION_ACCESS_VIOLATION if using native backend). Is there some known reason why this might be and does this indicate I'm not doing this right (you don't seem to need to do this in e.g. the compgraph)?

I realize I didn't manage to add a proper question about that "background work" kinda thing (slowdown was just motivation as to why I'm considering it): 

Putting the migrate call and operation in a new thread seems like it could work. Question is if there is a risk that the arrays get corrupted due to this (something which might be very hard to detect)? 

I guess a computer should be able to handle that several processes wants to move data between different places (e.g. RAM<->GPU) but things are often a fair bit more complicated than what that type of naive reasoning can shed light on.

**Comment by AlexDBlack** [comment]
> Is there some known reason why this might be and does this indicate I'm not doing this right

That's usually evidence of using an array out of scope - i.e., using an array after the workspace has been closed, or using an array from an old cycle (i.e., open workspace, allocate array X, close workspace, open workspace, use array X).
We've got pretty robust checks for using out of scope arrays (you'll get a Java exception) but maybe you're doing something that we don't have checks for.

> Putting the migrate call and operation in a new thread seems like it could work. Question is if there is a risk that the arrays get corrupted due to this (something which might be very hard to detect)?

Keep in mind workspaces are on a per-thread basis. And if you're migrating between threads on the CUDA background, you must call Nd4j.getExecutioner().commit() beforehand.

**Comment by DrChainsaw** [comment]
> open workspace, allocate array X, close workspace, open workspace, use array X

Ah, ok. This was probably the issue then.

I did this: "open workspace X, migrate activation for layer, close workspace X". 
Then when epsilon callback was made i did: "open workspace X, epsilon.migrate(false).muli(activation).amean(...), close workspace X".

Apparently it worked out to just store a reference to activation (without migrating it and without opening ws) and then do the epsilon callback handling just as above. 

When I did this it works without creating a new workspace name and I also got a handsome performance boost to the point where the slowdown is much more acceptable (order of 10-30%).

I expected an error for using leaked ws-pointer to activation or something similar. I guess I could just as well have been outta luck here depending on when memory for activations is reclaimed...

Thanks for the tip about commit. I'll try to keep it in mind should I ever need to try something with threads though

**Comment by AlexDBlack** [comment]
> I expected an error for using leaked ws-pointer to activation or something similar.

Yeah, we've got checks/exceptions for most cases for this (I thought it was pretty much all cases at this point, but apparently not).
How was the array used? i.e., do you know what method was being called on the out-of-scope array, or what op was using it?

**Comment by DrChainsaw** [comment]
What I meant was that just grabbing the activation (from the TrainingListener API), storing a reference to it (without detaching) and then multiplying the (migrated) epsilons with it seems like something which would generate a workspace panic exception, but it doesn't. I guess I'm just lucky that the activation is not out of scope when the epsilons are propagated because everything seems to work just fine now.

In case you were curious about what to triggered the exception before I refactored: I could not replicate the error in my refactored code so I would need to step back a revision or two. The only ops done was this though:  epsilon.migrate(false).muli(activation).amean(...). See my earlier post for rough workspace handling around this (although I suspect there was something more to it since I could not replicate).

Anyways, is there a way to persistently store an INDArray in a workspace? It seems like if the array is considered leaked after the ws goes out of scope then there is no way, or?

**Comment by AlexDBlack** [comment]
> epsilon.migrate(false).muli(activation).amean(...).

Thanks, I'll see if I can reproduce from that sequence of ops.

> is there a way to persistently store an INDArray in a workspace? It seems like if the array is considered leaked after the ws goes out of scope then there is no way, or?

The whole point of workspaces is for cyclical workloads.
There's really no point using a workspace for "persistent" (long-lived) arrays, where long-lived means more than one workspace cycle. So either keep the workspace open for as long as you need (might be a problem with some of the fit() workspace validation checks?), or don't use a workspace for really long-lived arrays.

Anyway, I'll close this issue (but I'll still get notified if you have more questions).
